### PR TITLE
Xdr add context standards

### PR DESCRIPTION
**Pull request by reutshal**
## Status
- [] In Progress
- [X] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/27965

## Description
- Added the following context outputs:
    - ***xdr-get-incident-extra-data*** command - Added *File*, *Process*, *IP* and *Domain* data.
    - ***xdr-get-audit-agent-reports*** command - Added *Endpoint* data.
    - ***xdr-get-endpoints*** command - Added *Account* data.

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [X] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.yml [hunk 1]
```diff
@@ -34,19 +34,18 @@ configuration:
...
-    seconds).
-  defaultvalue: 120
+- additionalinfo: The timeout of the HTTP requests sent to Cortex XDR API (in seconds).
+  defaultvalue: '120'
+  display: HTTP Timeout
   name: timeout
   required: false
   type: 0
-- defaultvalue: '10'
+- additionalinfo: The maximum number of incidents per fetch cannot exceed 100.
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.yml:43]:
```suggestion
- additionalinfo: The maximum number of incidents per fetch. cannot exceed 100.
```

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.yml [hunk 2]
```diff
@@ -535,9 +547,52 @@ script:
...
+      description: The full file name (including file extension).
+      type: String
+    - contextPath: Process.PID
+      description: The full file name (including file extension).
+      type: String
+    - contextPath: Process.Path
+      description: The full file name (including file extension).
+      type: String
+    - contextPath: Process.Start Time
+      description: The full file name (including file extension).
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.yml:578]: fix descriptions

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 3]
```diff
@@ -2751,7 +2905,7 @@ def main():
             raise
 
         demisto.error(traceback.format_exc())
-        return_error(str(err))
+        return_error(f'Error: {str(err)}\nTrace: \n{traceback.format_exc()}')
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:2908]: revert, no need to print the trace in the war-room

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 4]
```diff
@@ -1743,10 +1894,13 @@ def get_audit_agent_reports_command(client, args):
...
     )
 
+    endpoint_context = create_endpoint_context(audit_logs)
+
     return (
         tableToMarkdown('Audit Agent Reports', audit_logs),
         {
-            f'{INTEGRATION_CONTEXT_BRAND}.AuditAgentReports': audit_logs
+            f'{INTEGRATION_CONTEXT_BRAND}.AuditAgentReports': audit_logs,
+            Common.Endpoint.CONTEXT_PATH: endpoint_context
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:1903]:
```suggestion
            Common.Endpoint.CONTEXT_PATH: endpoint_context,
```

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 5]
```diff
@@ -1695,6 +1831,21 @@ def get_audit_management_logs_command(client, args):
     )
 
 
+def create_endpoint_context(audit_logs):
+    endpoints = []
+    for log in audit_logs:
+        endpoint_details = {
+            'ID': log.get('ENDPOINTID'),
+            'Hostname': log.get('ENDPOINTNAME'),
+            'Domain': log.get('DOMAIN')
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:1840]: [PEP8](https://www.python.org/dev/peps/pep-0008/#when-to-use-trailing-commas)
```suggestion
            'Domain': log.get('DOMAIN'),
```

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 6]
```diff
@@ -1412,10 +1542,16 @@ def get_endpoints_command(client, args):
             sort_by_first_seen=sort_by_first_seen,
             sort_by_last_seen=sort_by_last_seen
         )
+    context = {
+        f'{INTEGRATION_CONTEXT_BRAND}.Endpoint(val.endpoint_id == obj.endpoint_id)': endpoints,
+        'Endpoint(val.ID == obj.ID)': return_endpoint_standard_context(endpoints)
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:1547]:
```suggestion
        Common.Endpoint.CONTEXT_PATH: return_endpoint_standard_context(endpoints),
```

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 7]
```diff
@@ -1412,10 +1542,16 @@ def get_endpoints_command(client, args):
...
+    }
+    account_context = create_account_context(endpoints)
+    if account_context:
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:1550]: why do you have a check here and not on [here (line 1903)](https://github.com/demisto/content/pull/10627/files#diff-d3ec87f48a386d9632301be52a63780241c5103d4d43e80919e048c26cdd454eR1903)

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 8]
```diff
@@ -1695,6 +1831,21 @@ def get_audit_management_logs_command(client, args):
     )
 
 
+def create_endpoint_context(audit_logs):
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:1834]: move function to the top (helper function)

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 9]
```diff
@@ -1346,6 +1462,20 @@ def arg_to_int(arg, arg_name: str, required: bool = False):
     return ValueError(f'Invalid number: "{arg_name}"')
 
 
+def create_account_context(endpoints):
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:1465]: move function to the top (helper function)

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 10]
```diff
@@ -1346,6 +1462,20 @@ def arg_to_int(arg, arg_name: str, required: bool = False):
...
+    account_context = []
+    for endpoint in endpoints:
+        domain = endpoint.get('domain')
+        if domain:
+            for user in endpoint.get('users', []):
+                account_context.append({
+                    'Username': user,
+                    'Domain': domain
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:1473]:
```suggestion
                    'Domain': domain,
```

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 11]
```diff
@@ -1208,6 +1208,111 @@ def get_incidents_command(client, args):
...
+                ip_context.append(network_ip_details)
+
+    return domain_context
+
+
+def get_indicators_context(incident):
+    file_context: List[Any] = []
+    process_context: List[Any] = []
+    ip_context: List[Any] = []
+    for alert in incident.get('alerts'):
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:1279]:
```suggestion
    for alert in incident.get('alerts', []):
```

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 12]
```diff
@@ -1208,6 +1208,111 @@ def get_incidents_command(client, args):
...
+        # file context
+        file_details = {
+            'Name': alert.get('action_file_name'),
+            'Path': alert.get('action_file_path'),
+            'SHA265': alert.get('action_file_sha256'),
+            'MD5': alert.get('action_file_md5')
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:1285]:
```suggestion
            'MD5': alert.get('action_file_md5'),
```

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 13]
```diff
@@ -1208,6 +1208,111 @@ def get_incidents_command(client, args):
...
+
+    network_artifacts = incident.get('network_artifacts', [])
+
+    domain_context = create_context_from_network_artifacts(network_artifacts, ip_context)
+
+    file_artifacts = incident.get('file_artifacts', [])
+    for file in file_artifacts:
+        file_context.append({
+            'Name': file.get('file_name'),
+            'SHA256': file.get('file_sha256')
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:1310]:
```suggestion
            'SHA256': file.get('file_sha256'),
```

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 14]
```diff
@@ -1208,6 +1208,111 @@ def get_incidents_command(client, args):
...
+        # ip context
+        add_to_ip_context(alert, ip_context)
+
+    network_artifacts = incident.get('network_artifacts', [])
+
+    domain_context = create_context_from_network_artifacts(network_artifacts, ip_context)
+
+    file_artifacts = incident.get('file_artifacts', [])
+    for file in file_artifacts:
+        file_context.append({
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:1308]: why there isn't a validation on the value before appending to the list as you do in the previous loop?

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 15]
```diff
@@ -1208,6 +1208,111 @@ def get_incidents_command(client, args):
...
+def create_context_from_network_artifacts(network_artifacts, ip_context):
+    domain_context = []
+
+    if network_artifacts:
+        for artifact in network_artifacts:
+            domain = artifact.get('network_domain')
+            if domain:
+                domain_context.append(
+                    {
+                        'Name': domain
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:1258]:
```suggestion
                        'Name': domain,
```

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 16]
```diff
@@ -1208,6 +1208,111 @@ def get_incidents_command(client, args):
...
+                    }
+                )
+
+            network_ip_details = {
+                'Address': artifact.get('network_remote_ip'),
+                'GEO.Country': artifact.get('network_country')
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:1264]: make sure it parses correctly in the context and not being exported as a context key with a dot (I think it should be fine).
```suggestion
                'GEO.Country': artifact.get('network_country'),
```

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 17]
```diff
@@ -1208,6 +1208,111 @@ def get_incidents_command(client, args):
...
+
+def add_to_ip_context(alert, ip_context):
+    action_local_ip = alert.get('action_local_ip')
+    action_remote_ip = alert.get('action_remote_ip')
+    if action_local_ip:
+        ip_context.append(
+            {
+                'Address': action_local_ip
+            }
+        )
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:1239]: no reason to have each bracket in a separate line
```suggestion
        ip_context.append({
                'Address': action_local_ip,
        })
```

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 18]
```diff
@@ -1208,6 +1208,111 @@ def get_incidents_command(client, args):
...
+
+    if action_remote_ip:
+        ip_context.append(
+            {
+                'Address': action_remote_ip
+            }
+        )
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:1246]:
```suggestion
        ip_context.append({
                'Address': action_remote_ip,
        })
```

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py [hunk 19]
```diff
@@ -1208,6 +1208,111 @@ def get_incidents_command(client, args):
...
+def get_process_context(alert, process_type):
+    process_context = {
+        'Name': alert.get(f'{process_type}_process_image_name'),
+        'MD5': alert.get(f'{process_type}_process_image_md5'),
+        'SHA256': alert.get(f'{process_type}_process_image_sha256'),
+        'PID': alert.get(f'{process_type}_process_os_pid'),
+        'CommandLine': alert.get(f'{process_type}_process_command_line'),
+        'Path': alert.get(f'{process_type}_process_image_path'),
+        'Start Time': alert.get(f'{process_type}_process_execution_time'),
+        'Hostname': alert.get('host_name')
```
Comment (yaakovi) [inline Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:1220]:
```suggestion
        'Hostname': alert.get('host_name'),
```

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py (hunk 3 above)
Comment (reutshal) [reply Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:2908]: Why not? When you have an error in one of your methods it helps to know where.
It solved an issue for me during the development here

**Comment by reutshal** [comment]
@roysagi Can you please go over the followings:
* Yml file - Only the added outputs:
https://github.com/demisto/content/pull/10627/files#diff-2707e061000ea8be0033e324685f9a60cde07cee318eb83b2764ffd79fd8e8b5R550
https://github.com/demisto/content/pull/10627/files#diff-2707e061000ea8be0033e324685f9a60cde07cee318eb83b2764ffd79fd8e8b5R1020
https://github.com/demisto/content/pull/10627/files#diff-2707e061000ea8be0033e324685f9a60cde07cee318eb83b2764ffd79fd8e8b5R1522
* RN: https://github.com/demisto/content/blob/XDR-Add_context_standads/Packs/CortexXDR/ReleaseNotes/2_7_2.md
Thanks!

File: Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py (hunk 3 above)
Comment (yaakovi) [reply Packs/CortexXDR/Integrations/CortexXDRIR/CortexXDRIR.py:2904]: the traceback will be in the server log, there is no reason to print it in the war-room. it's not very esthetic.

**Review by yaakovi — APPROVED** [review]

**Pull request merged**